### PR TITLE
[SPARK-42286][SQL] Fallback to previous codegen code path for complex expr with CAST

### PR DESCRIPTION
**Pull request by RunyaoChen**
### What changes were proposed in this pull request?

This PR fixes the internal error `Child is not Cast or ExpressionProxy of Cast` for valid `CaseWhen` expr with `Cast` expr in its branches.

Specifically, after SPARK-39865, an improved error msg for overflow exception during table insert was introduced. The improvement covers `Cast` expr and `ExpressionProxy` expr, but `CaseWhen` and other complex ones are not covered. An example below hits an internal error today.
```
create table t1 as select x FROM values (1), (2), (3) as tab(x);
create table t2 (x Decimal(9, 0));
insert into t2 select 0 - (case when x = 1 then 1 else x end) from t1 where x = 1;
```
To fix the query failure, we decide to fall back to the previous handling if the expr is not a simple `Cast` expr or `ExpressionProxy` expr.

### Why are the changes needed?

To fix the query regression introduced in SPARK-39865.

### Does this PR introduce _any_ user-facing change?
No. We just fall back to the previous error msg if the expression involving `Cast` is not a simple one.

### How was this patch tested?
- Added Unit test.
- Removed one test case for the `Child is not Cast or ExpressionProxy of Cast` internal error, as now we do not check if the child has a `Cast` expression and fall back to the previous error message.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 1]
```diff
@@ -2511,40 +2511,41 @@ case class UpCast(child: Expression, target: AbstractDataType, walkedTypePath: S
...
   }
 
-  private def getCast: Cast = child match {
+  private def getCast: Option[Cast] = child match {
     case c: Cast =>
-      c
+      Some(c)
     case ExpressionProxy(c, _, _) =>
-      c.asInstanceOf[Cast]
+      Some(c.asInstanceOf[Cast])
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:2523]: How about 
```
case ExpressionProxy(c: Cast, _, _) =>
      Some(c)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 2]
```diff
@@ -2511,40 +2511,41 @@ case class UpCast(child: Expression, target: AbstractDataType, walkedTypePath: S
...
 
   override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
-    val child = getCast
+    getCast match {
+      case Some(child) => doGenCodeWithNewErrorMsg(ctx, ev, child)
+      case None => child.genCode(ctx)
+    }
+  }
+
+  def doGenCodeWithNewErrorMsg(ctx: CodegenContext, ev: ExprCode, child: Cast): ExprCode = {
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:2548]: How about `doGenCodeWithBetterErrorMsg`?

**Comment by gengliangwang** [comment]
cc @bersprockets as well

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala (hunk 1 above)
Comment (RunyaoChen) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: Done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala (hunk 2 above)
Comment (RunyaoChen) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: Done.

**Comment by bersprockets** [comment]
Just a couple of questions:

>create or replace table es570639t1 as select x FROM values (1), (2), (3) as tab(x);
>create or replace table es570639t2 (x Decimal(9, 0));

For the eventual commit message: Should that be `create table`.... instead of `create or replace table`...? or is that new syntax?

For my own understanding: it seems that `CheckOverflowInTableInsert` previously assumed a `Cast` as child (and after SPARK-41991, assumed either a `Cast` or one wrapped by an `ExpressionProxy` ), but rules like `PushFoldableIntoBranches` make that assumption too rigid, correct?

**Comment by RunyaoChen** [comment]
>For the eventual commit message: Should that be create table.... instead of create or replace table...? or is that new syntax?

Create table should suffice to repo this issue. I updated the commit msg. Thanks for spotting this!

**Comment by RunyaoChen** [comment]
> For my own understanding: it seems that CheckOverflowInTableInsert previously assumed a Cast as child (and after [SPARK-41991](https://issues.apache.org/jira/browse/SPARK-41991), assumed either a Cast or one wrapped by an ExpressionProxy ), but rules like PushFoldableIntoBranches make that assumption too rigid, correct?

Yes, rules like `PushFoldableIntoBranches` push the `Cast` down to each branch of `CaseWhen`, and what `checkChild` sees is a `CaseWhen` expr with `Cast` in each branch (in this case, it sees `CASE WHEN (x#4 = 1) THEN cast(-1 as decimal(9,0)) ELSE cast((0 - x#4) as decimal(9,0)) END`).

**Comment by gengliangwang** [comment]
@RunyaoChen shall we add an end-to-end test for this one? I will merge it after that.

**Comment by RunyaoChen** [comment]
> @RunyaoChen shall we add an end-to-end test for this one? I will merge it after that.

Sure, added an INSERT INTO query with SELECT Case When that throws the previous CAST_OVERFLOW exception.

**Comment by RunyaoChen** [comment]
> @RunyaoChen shall we add an end-to-end test for this one? I will merge it after that.

@gengliangwang added positive case as well

**Review by gengliangwang — APPROVED** [review]

**Comment by gengliangwang** [comment]
Thanks, merging to master/3.4

**Comment by wangyum** [comment]
@RunyaoChen Could you backport this fix to branch-3.3 to fix SPARK-42473?

**Comment by RunyaoChen** [comment]
> @RunyaoChen Could you backport this fix to branch-3.3 to fix [SPARK-42473](https://issues.apache.org/jira/browse/SPARK-42473)?

Sure, here's the cherry-pick to branch-3.3: https://github.com/apache/spark/pull/40140
cc: @gengliangwang

**Pull request closed without merging**